Add tan_cuda for complex dtypes

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38400 Add tan_cuda for complex dtypes**

Differential Revision: [D21572209](https://our.internmc.facebook.com/intern/diff/D21572209)

anjali411 committed:
Add tan_cuda for complex dtypes

[ghstack-poisoned]

anjali411 committed:
Update on "Add tan_cuda for complex dtypes"




[ghstack-poisoned]

albanD commented on test/test_torch.py:
```diff
@@ -17713,6 +17715,7 @@ def inner(self, device, dtype):
     ('rsqrt', '', lambda t, d: _small_3d(t, d) + 1, lambda t, d: [], 1e-2, 1e-5, 1e-4, _float_types_no_half),
     ('sinh', '', lambda t, d: _small_3d(t, d).clamp(-1, 1), lambda t, d: [], 1e-3, 1e-5, 1e-5, _float_types),
     ('tan', '', lambda t, d: _small_3d(t, d).clamp(-1, 1), lambda t, d: [], 1e-3, 1e-5, 1e-5, _float_types),
+    ('tan', 'complex', lambda t, d: _small_3d(t, d), lambda t, d: [], 1e-3, 1e-5, 1e-5, _complex_types)
```
I am surprised that `_complex_types` did not exist yet here. How do we test the other methods that already support complex?

anjali411 replied:
Unfortunately there are no existing tests for these ops for complex dtypes. we should add it soon though ...

albanD commented on test/test_autograd.py:
```diff
@@ -4185,7 +4185,7 @@ def run_functional_checks(test_case, test_name, name, apply_fn, run_grad_checks,
         test_case.assertEqual(self_variable.size(), self_variable.grad.size())
 
-separate_complex_tests = ['log', 'log10', 'log1p', 'log2', 'reciprocal']
+separate_complex_tests = ['log', 'log10', 'log1p', 'log2', 'reciprocal', 'tan']
```
Why is this one in the reparate list? It should work for real numbers as well no?

anjali411 replied:
the test for floating points uses clamp in the input and clamp does not work for complex

anjali411 committed:
Update on "Add tan_cuda for complex dtypes"




[ghstack-poisoned]

albanD reviewed (approved):
LGTM

anjali411 committed:
Update on "Add tan_cuda for complex dtypes"




[ghstack-poisoned]

anjali411 committed:
Update on "Add tan_cuda for complex dtypes"




[ghstack-poisoned]

anjali411 committed:
Update on "Add tan_cuda for complex dtypes"


Differential Revision: [D21556867](https://our.internmc.facebook.com/intern/diff/D21556867)

[ghstack-poisoned]

anjali411 committed:
Update on "Add tan_cuda for complex dtypes"

* #38399 Added autograd tests, disabled jit autograd tests for complex and added a separate list for tests for complex dtype only

[ghstack-poisoned]

anjali411 committed:
Update on "Add tan_cuda for complex dtypes"


Differential Revision: [D21572209](https://our.internmc.facebook.com/intern/diff/D21572209)

[ghstack-poisoned]
